content/blog/2022-05-08-Apache-ShenYu(incubating)-Integrated-Skywalking-Practice-Observability/index.md

```diff
@@ -0,0 +1,130 @@
+
+
+## 3. Adding generalized call tracking to the gRPC plugin and keeping it compatible
+
+The existing SkyWalking gRPC plugin only supports calls initiated by way of stubs. For the gateway there is no proto file, the gateway takes generalized calls (not through stubs), so tracing rpc requests, you will find that the link will break at the gateway node.
+In this case, it is necessary to make the gRPC plugin support generalized calls, while at the same time needing to remain compatible and not affect the original tracing method. This is achieved by determining whether the request parameter is a DynamicMessage, and if it is not, then the original tracing logic through the stub is used.
+If not, then the original tracing logic via stubs is used, and if not, then the generalized call tracing logic is used. The other compatibility is the difference between the old and new versions of grpc, as well as the compatibility of various cases of obtaining server-side IP, for those interested in the source code.
+
+![](grpc-generic-call.png)
+
+## 4. ShenYu Gateway Observability Practice
+The above explains the principle of SkyWalking ShenYu plug-in implementation, the following deployment of applications to see the effect. SkyWalking powerful, in addition to the link tracking requires the development of plug-ins, other powerful features out of the box.
+Here only describe the link tracking and application performance analysis part, if you want to experience the power of SkyWalking features, please refer to the [SkyWalking official documentation](https://skywalking.apache.org/)
+
+### 4.1 Deploying Applications
+
+- After packaging SkyWalking Agent (` . /mvnw clean package -Pall -DskipTests`, recommended to download directly from [official website](https://skywalking.apache.org/downloads/) ), for the tracking of the shenyu gateway, put the shenyu plugin into the plugins directory
+- For the service to be tracked, add the following parameter to the start jar command: `-javaagent:skywalking-agent.jar's path -DSW_AGENT_NAME=Your_ApplicationName`
+- Local environment experience the effect can be directly used H2 database, out of the box, the production environment according to the specific changes agent.config
+
+### 4.2 Sending requests to the gateway
+Initiate various services to the gateway through the postman client.
+
+### 4.3 Request Topology Diagram
+![](topology.png)
+
+### 4.4 Request Trace (in the case of gRPC)
+
+Normal Trace：  
+![](grpc-ok.png)
+
+Abnormal Trace：  
+![](grpc-error.png)
+
+Click on the link node to see the corresponding node information and exception information   
+Service Provider Span  
+![](grpc-service-span.png)
+
+Gateway request span
+![](gateway-error-sapn.png)
+
+### 4.5 Service Metrics Monitoring
+
+Service Metrics Monitoring
+![](overview.png)
+
+### 4.6 Database metrics monitoring
+![](database.png)
+
+### 4.7 JVM Monitoring
+![](jvm.png)
```

Comment:
Do you consider a thread pool monitoring?

Reply:
Thread pooling and connection pooling are of great concern to everyone, this is illustrated in the gateway management background with a diagram, which also reflects the power of SkyWalking